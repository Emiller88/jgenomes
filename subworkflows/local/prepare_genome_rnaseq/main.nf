include { BOWTIE2_BUILD                                         } from '../../../modules/nf-core/bowtie2/build'
include { BOWTIE_BUILD as BOWTIE1_BUILD                         } from '../../../modules/nf-core/bowtie/build'
include { GFFREAD                                               } from '../../../modules/nf-core/gffread'
include { HISAT2_BUILD                                          } from '../../../modules/nf-core/hisat2/build'
include { HISAT2_EXTRACTSPLICESITES                             } from '../../../modules/nf-core/hisat2/extractsplicesites'
include { KALLISTO_INDEX                                        } from '../../../modules/nf-core/kallisto/index'
include { RSEM_PREPAREREFERENCE as MAKE_TRANSCRIPTS_FASTA       } from '../../../modules/nf-core/rsem/preparereference'
include { RSEM_PREPAREREFERENCE as RSEM_PREPAREREFERENCE_GENOME } from '../../../modules/nf-core/rsem/preparereference'
include { SALMON_INDEX                                          } from '../../../modules/nf-core/salmon/index'
include { SAMTOOLS_FAIDX                                        } from '../../../modules/nf-core/samtools/faidx'
include { STAR_GENOMEGENERATE                                   } from '../../../modules/nf-core/star/genomegenerate'

workflow PREPARE_GENOME_RNASEQ {
    take:
    fasta                          // channel: [meta, fasta]
    fasta_fai                      // channel: [meta, fasta_fai]
    gff                            // channel: [meta, gff]
    gtf                            // channel: [meta, gtf]
    splice_sites                   // channel: [meta, splice_sites]
    transcript_fasta               // channel: [meta, transcript_fasta]
    run_bowtie1                    // boolean: true/false
    run_bowtie2                    // boolean: true/false
    run_faidx                      // boolean: true/false
    run_hisat2                     // boolean: true/false
    run_hisat2_extractsplicesites  // boolean: true/false
    run_kallisto                   // boolean: true/false
    run_rsem                       // boolean: true/false
    run_rsem_make_transcript_fasta // boolean: true/false
    run_salmon                     // boolean: true/false
    run_sizes                      // boolean: true/false
    run_star                       // boolean: true/false

    main:
    bowtie1_index = Channel.empty()
    bowtie2_index = Channel.empty()
    hisat2_index = Channel.empty()
    kallisto_index = Channel.empty()
    rsem_index = Channel.empty()
    salmon_index = Channel.empty()
    star_index = Channel.empty()
    fasta_sizes = Channel.empty()

    versions = Channel.empty()

    if (run_bowtie1) {
        BOWTIE1_BUILD(fasta)

        bowtie1_index = BOWTIE1_BUILD.out.index
        versions = versions.mix(BOWTIE1_BUILD.out.versions)
    }

    if (run_bowtie2) {
        BOWTIE2_BUILD(fasta)

        bowtie2_index = BOWTIE2_BUILD.out.index
        versions = versions.mix(BOWTIE2_BUILD.out.versions)
    }

    if (run_faidx || run_sizes) {
        SAMTOOLS_FAIDX(
            fasta,
            [[id: 'no_fai'], []],
            run_sizes,
        )

        fasta_fai = fasta_fai.mix(SAMTOOLS_FAIDX.out.fai)
        fasta_sizes = SAMTOOLS_FAIDX.out.sizes
        versions = versions.mix(SAMTOOLS_FAIDX.out.versions)
    }

    if (run_hisat2 || run_kallisto || run_rsem || run_rsem_make_transcript_fasta || run_salmon || run_star) {
        GFFREAD(
            gff,
            [],
        )

        versions = versions.mix(GFFREAD.out.versions)

        gtf = gtf
            .mix(GFFREAD.out.gtf)
            .groupTuple()
            .map { meta, gtf_ -> gtf_[1] ? [meta, gtf_[1]] : [meta, gtf_] }

        if (run_hisat2 || run_hisat2_extractsplicesites) {
            HISAT2_EXTRACTSPLICESITES(gtf)

            versions = versions.mix(HISAT2_EXTRACTSPLICESITES.out.versions)
            splice_sites = splice_sites.mix(HISAT2_EXTRACTSPLICESITES.out.txt)

            if (run_hisat2) {
                HISAT2_BUILD(
                    fasta,
                    gtf,
                    splice_sites,
                )

                hisat2_index = HISAT2_BUILD.out.index

                versions = versions.mix(HISAT2_BUILD.out.versions)
            }
        }

        if (run_kallisto || run_rsem_make_transcript_fasta || run_salmon) {
            MAKE_TRANSCRIPTS_FASTA(
                fasta,
                gtf,
            )
            versions = versions.mix(MAKE_TRANSCRIPTS_FASTA.out.versions)

            transcript_fasta = transcript_fasta.mix(MAKE_TRANSCRIPTS_FASTA.out.transcript_fasta)

            if (run_kallisto) {
                KALLISTO_INDEX(transcript_fasta)

                kallisto_index = KALLISTO_INDEX.out.index
                versions = versions.mix(KALLISTO_INDEX.out.versions)
            }

            if (run_salmon) {
                SALMON_INDEX(
                    fasta,
                    transcript_fasta,
                )

                salmon_index = SALMON_INDEX.out.index
                versions = versions.mix(SALMON_INDEX.out.versions)
            }
        }

        if (run_rsem) {
            RSEM_PREPAREREFERENCE_GENOME(fasta, gtf)

            rsem_index = RSEM_PREPAREREFERENCE_GENOME.out.index
            versions = versions.mix(RSEM_PREPAREREFERENCE_GENOME.out.versions)
        }

        if (run_star) {
            STAR_GENOMEGENERATE(fasta, gtf)

            star_index = STAR_GENOMEGENERATE.out.index
            versions = versions.mix(STAR_GENOMEGENERATE.out.versions)
        }
    }

    emit:
    bowtie1_index    // channel: [meta, BowtieIndex/]
    bowtie2_index    // channel: [meta, Bowtie2Index/]
    fasta_fai        // channel: [meta, *.fa(sta).fai]
    fasta_sizes      // channel: [meta, *.fa(sta).sizes]
    gtf              // channel: [meta, gtf]
    hisat2_index     // channel: [meta, Hisat2Index/]
    kallisto_index   // channel: [meta, KallistoIndex]
    rsem_index       // channel: [meta, RSEMIndex/]
    salmon_index     // channel: [meta, SalmonIndex/]
    splice_sites     // channel: [meta, *.splice_sites.txt]
    star_index       // channel: [meta, STARIndex/]
    transcript_fasta // channel: [meta, *.transcripts.fasta]
    versions         // channel: [versions.yml]
}
