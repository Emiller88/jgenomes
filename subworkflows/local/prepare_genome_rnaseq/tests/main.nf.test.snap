{
    ".fasta/.gff - all tools": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "test.1.ebwt:md5,e728d922a147241f699fdad1f7ac5a9d",
                        "test.2.ebwt:md5,c0a8d0fffdd7e7a3d9229e9d42a7964a",
                        "test.3.ebwt:md5,123b1717a761500ffa59d54959215d26",
                        "test.4.ebwt:md5,7c9d42e29a60c13a65dd0046a012d848",
                        "test.rev.1.ebwt:md5,dbcbf4caa51963abb964bd59064aeffa",
                        "test.rev.2.ebwt:md5,7cfa0c2adf3488fd184008a7346abba7"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "genome.1.bt2:md5,118f83ac94277f7058d1ecb79fb5acf9",
                        "genome.2.bt2:md5,10a474242f1d77979e1dfc3d862faac8",
                        "genome.3.bt2:md5,123b1717a761500ffa59d54959215d26",
                        "genome.4.bt2:md5,7c9d42e29a60c13a65dd0046a012d848",
                        "genome.rev.1.bt2:md5,21f198b3402e4119a181982ee8b4efd3",
                        "genome.rev.2.bt2:md5,352e3ab4c780f58a7c97f1f6978949a8"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "genome.fasta.fai:md5,3520cd30e1b100e55f578db9c855f685"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "genome.fasta.sizes:md5,b190587cae0531f3cf25552d8aa674db"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "test.gtf:md5,eb77931882d554f6fcdb2ebbfec062ee"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "genome.1.ht2:md5,8f8a9b1b43511fbc679454cc6f93d203",
                        "genome.2.ht2:md5,afe55bf628b41121cf2c9a999e1dbbcb",
                        "genome.3.ht2:md5,123b1717a761500ffa59d54959215d26",
                        "genome.4.ht2:md5,7c9d42e29a60c13a65dd0046a012d848",
                        "genome.5.ht2:md5,0c5bdeb9777a0bb5c2da117c7b0e2a85",
                        "genome.6.ht2:md5,e055ccdcad39cf608d234ab650eabbb2",
                        "genome.7.ht2:md5,b9e300d0e85cdde63bd6fdcc0d5d5097",
                        "genome.8.ht2:md5,6ae6e44dc59b540e1caadb5b8eeb394a"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "kallisto:md5,89b07a5fbea444b814c1a42064dbb8c5"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "Genome:md5,a654229fbca6071dcb6b01ce7df704da",
                        "Log.out:md5,31a3a433773f31fc9e44a7d722861826",
                        "SA:md5,8c3edc46697b72c9e92440d4cf43506c",
                        "SAindex:md5,fd05c149960e72642a8d7c860528ae81",
                        "chrLength.txt:md5,c81f40f27e72606d7d07097c1d56a5b5",
                        "chrName.txt:md5,5ae68a67b70976ee95342a7451cb5af1",
                        "chrNameLength.txt:md5,b190587cae0531f3cf25552d8aa674db",
                        "chrStart.txt:md5,8d3291e6bcdbe9902fbd7c887494173f",
                        "exonGeTrInfo.tab:md5,72d4dd88d25e3c5b0bb72b12b4ac99a8",
                        "exonInfo.tab:md5,0d560290fab688b7268d88d5494bf9fe",
                        "geneInfo.tab:md5,8b608537307443ffaee4927d2b428805",
                        "genome.chrlist:md5,b190587cae0531f3cf25552d8aa674db",
                        "genome.fasta:md5,f315020d899597c1b57e5fe9f60f4c3e",
                        "genome.grp:md5,c2848a8b6d495956c11ec53efc1de67e",
                        "genome.idx.fa:md5,050c521a2719c2ae48267c1e65218f29",
                        "genome.n2g.idx.fa:md5,050c521a2719c2ae48267c1e65218f29",
                        "genome.seq:md5,94da0c6b88c33e63c9a052a11f4f57c1",
                        "genome.ti:md5,b0815127b7a48f82f705fd35e1bdf528",
                        "genome.transcripts.fa:md5,050c521a2719c2ae48267c1e65218f29",
                        "genomeParameters.txt:md5,6f7d4d17e53630a7f805605174b5ea4d",
                        "sjdbInfo.txt:md5,5690ea9d9f09f7ff85b7fd47bd234903",
                        "sjdbList.fromGTF.out.tab:md5,8760c33e966dad0b39f440301ebbdee4",
                        "sjdbList.out.tab:md5,9e4f991abbbfeb3935a2bb21b9e258f1",
                        "transcriptInfo.tab:md5,0c3a5adb49d15e5feff81db8e29f2e36"
                    ]
                ]
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.splice_sites.txt:md5,dc6da761e1dda4b7c490eea01260ee21"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "Genome:md5,a654229fbca6071dcb6b01ce7df704da",
                        "Log.out:md5,1414afc9e7d5140cb1d4666e597919d0",
                        "SA:md5,8c3edc46697b72c9e92440d4cf43506c",
                        "SAindex:md5,d0fbf2789ee1e9f60c352ba3655d9de4",
                        "chrLength.txt:md5,c81f40f27e72606d7d07097c1d56a5b5",
                        "chrName.txt:md5,5ae68a67b70976ee95342a7451cb5af1",
                        "chrNameLength.txt:md5,b190587cae0531f3cf25552d8aa674db",
                        "chrStart.txt:md5,8d3291e6bcdbe9902fbd7c887494173f",
                        "exonGeTrInfo.tab:md5,72d4dd88d25e3c5b0bb72b12b4ac99a8",
                        "exonInfo.tab:md5,0d560290fab688b7268d88d5494bf9fe",
                        "geneInfo.tab:md5,8b608537307443ffaee4927d2b428805",
                        "genomeParameters.txt:md5,88b05e908bf56cfc04cf85abec045be9",
                        "sjdbInfo.txt:md5,5690ea9d9f09f7ff85b7fd47bd234903",
                        "sjdbList.fromGTF.out.tab:md5,8760c33e966dad0b39f440301ebbdee4",
                        "sjdbList.out.tab:md5,9e4f991abbbfeb3935a2bb21b9e258f1",
                        "transcriptInfo.tab:md5,0c3a5adb49d15e5feff81db8e29f2e36"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "genome.transcripts.fa:md5,050c521a2719c2ae48267c1e65218f29"
                ]
            ],
            [
                "versions.yml:md5,2e620478a8847b07ef43e539f3d15a4d",
                "versions.yml:md5,3ef2e2c4f68cff618c1ea7f506974348",
                "versions.yml:md5,8c52d613ef359e337bdfff8893022ff0",
                "versions.yml:md5,8c7e080728ac03ca7d12075b04269c87",
                "versions.yml:md5,9089b357f16db5a1a7f8c19a4a8d347f",
                "versions.yml:md5,9c05faacdd2f90677bfc81c7c70454e7",
                "versions.yml:md5,bdd0980428bddd82b00efe49c7e33203",
                "versions.yml:md5,c082767988a5e7ee97408d28358389aa",
                "versions.yml:md5,c4c0ac84f821dee47cbb5d876778e301",
                "versions.yml:md5,e8962da49ea2a0280fa1ff74caa494c5"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.01.0"
        },
        "timestamp": "2025-03-07T12:48:28.175203873"
    }
}